packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,293 @@
+  lastFocused?: HTMLElement,
+  onFocus: () => void,
+  onBlur: () => void
+};
+
+class LandmarkManager {
+  private landmarks: Array<Landmark> = [];
+  private static instance: LandmarkManager;
+
+  private constructor() {}
+
+  public static getInstance(): LandmarkManager {
+    if (!LandmarkManager.instance) {
+      LandmarkManager.instance = new LandmarkManager();
+      LandmarkManager.instance.setup();
+    }
+
+    return LandmarkManager.instance;
+  }
+
+  private setup() {
+    document.addEventListener('keydown', LandmarkManager.getInstance().f6Handler.bind(LandmarkManager.getInstance()), {capture: true});
+    document.addEventListener('focusin', LandmarkManager.getInstance().focusinHandler.bind(LandmarkManager.getInstance()), {capture: true});
+  }
+
+  private focusLandmark(landmark: HTMLElement) {
+    this.landmarks.find(l => l.ref.current === landmark)?.onFocus();
+  }
+
+  /**
+   * Return set of landmarks with a specific role.
+   */
+  public getLandmarksByRole(role: AriaLandmarkRole) {
+    return new Set(this.landmarks.filter(l => l.role === role));
+  }
+
+  /**
+   * Return first landmark with a specific role.
+   */
+  public getLandmarkByRole(role: AriaLandmarkRole) {
+    let landmarks = this.landmarks.filter(l => l.role === role);
+    if (landmarks.length > 0) {
+      return landmarks[0];
+    }
+  }
+
+  public addLandmark(newLandmark: Landmark) {
+    if (this.landmarks.find(landmark => landmark.ref === newLandmark.ref)) {
+      return;
+    }
+
+    if (this.landmarks.filter(landmark => landmark.role === 'main').length > 1) {
+      console.error('Page can contain no more than one landmark with the role "main".');
+    }
+
+    if (this.landmarks.length === 0) {
+      this.landmarks = [newLandmark];
+      return;
+    }
+
+    let insertPosition = 0;
+    let comparedPosition = newLandmark.ref.current.compareDocumentPosition(this.landmarks[insertPosition].ref.current as Node);
+      // Compare position of landmark being added with existing landmarks.
+      // Iterate through landmarks (which are sorted in document order),
+      // and insert when a landmark is found that is positioned before the newly added element,
+      // or is contained by the newly added element (for nested landmarks).
+      // https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition
+    while (
+        insertPosition < this.landmarks.length &&
+        ((comparedPosition & Node.DOCUMENT_POSITION_PRECEDING) ||
+        (comparedPosition & Node.DOCUMENT_POSITION_CONTAINS))
+        ) {
+      comparedPosition = newLandmark.ref.current.compareDocumentPosition(this.landmarks[insertPosition].ref.current as Node);
+      insertPosition++;
+    }
+    this.landmarks.splice(insertPosition, 0, newLandmark);
+  }
+
+  public updateLandmark(landmark: Landmark) {
+    this.landmarks = this.landmarks.map((prevLandmark) => prevLandmark.ref.current === landmark.ref.current ? {...prevLandmark, ...landmark} : prevLandmark);
```

Comment:
Do landmarks need to be immutable? Wouldn't it be easier to just update the properties on the existing one?